File: src/main/java/sharev/domain/team/repository/TeamRepositoryImpl.java
package sharev.domain.team.repository;

import static sharev.domain.account.entity.QAccount.account;
import static sharev.domain.member.entity.QMember.member;
import static sharev.domain.team.entity.QTeam.team;

import com.querydsl.jpa.JPAExpressions;
import com.querydsl.jpa.impl.JPAQueryFactory;
import java.util.List;
import lombok.RequiredArgsConstructor;
import sharev.domain.member.entity.QMember;
import sharev.domain.team.dto.response.QResponseTeamInfoDto;
import sharev.domain.team.dto.response.QTempTeamMemberInfoDto;
import sharev.domain.team.dto.response.ResponseTeamInfoDto;
import sharev.domain.team.dto.response.TempTeamMemberInfoDto;

@RequiredArgsConstructor
public class TeamRepositoryImpl implements TeamRepositoryCustom {
    private final JPAQueryFactory queryFactory;

    public List<ResponseTeamInfoDto> findMyTeams(Long accountId) {
        QMember teamMember = new QMember("teamMember");

        return queryFactory
                .select(new QResponseTeamInfoDto(
                        team.id,
                        team.title,
                        team.content,
                        team.createdAt,
                        member.role,
                        JPAExpressions
                                .select(teamMember.count().intValue())
                                .from(teamMember)
                                .where(teamMember.team.eq(team))
                ))
                .from(member)

                .join(team)
                .on(member.team.eq(team))

                .where(member.account.id.eq(accountId))
                .fetch();
    }

    public List<TempTeamMemberInfoDto> findMyTeamMembers(Long teamId) {
        return queryFactory
                .select(new QTempTeamMemberInfoDto(
                        account.name,
                        account.email
                ))
                .from(member)

                .leftJoin(account)
                .on(member.account.eq(account))

                .where(member.team.id.eq(teamId))
                .fetch();
    }
}

File: src/main/kotlin/sharev/team/adapter/outbound/jpa/repository/TeamRepositoryImpl.kt
package sharev.team.adapter.outbound.jpa.repository

import jakarta.persistence.EntityManager
import org.springframework.stereotype.Repository
import sharev.member.domain.model.MemberRole
import sharev.team.application.port.outbound.summary.TeamMemberSummary
import sharev.team.application.port.outbound.summary.TeamSummary
import java.time.LocalDateTime

@Repository
class TeamRepositoryImpl(
    private val entityManager: EntityManager,
) : TeamRepositoryCustom {

    override fun findMyTeams(accountId: Long): List<TeamSummary> {
        val rows = entityManager.createQuery(
            """
            select t.id, t.title, t.content, t.createdAt, m.role,
                   (select count(m2.id) from MemberJpaEntity m2 where m2.team = t)
            from MemberJpaEntity m
            join m.team t
            where m.account.id = :accountId
            """.trimIndent(),
            Array<Any>::class.java
        )
            .setParameter("accountId", accountId)
            .resultList

        return rows.map { row ->
            TeamSummary(
                id = row[0] as Long,
                title = row[1] as String,
                content = row[2] as String?,
                createdAt = row[3] as LocalDateTime?,
                memberRole = row[4] as String,
                headcount = (row[5] as Long).toInt(),
            )
        }
    }

    override fun findMyTeamMembers(teamId: Long): List<TeamMemberSummary> {
        val rows = entityManager.createQuery(
            """
            select m.account.name, m.account.email, m.role
            from MemberJpaEntity m
            where m.team.id = :teamId
            """.trimIndent(),
            Array<Any>::class.java
        )
            .setParameter("teamId", teamId)
            .resultList

        return rows.map { row ->
            TeamMemberSummary(
                name = row[0] as String,
                email = row[1] as String,
                role = row[2] as MemberRole,
            )
        }
    }
}
